### ISSUE TITLE
Research: CSV based bulk inserts

### ISSUE THREAD

**Issue by Grauenwolf** [docevaad/Chain#230]
https://stackoverflow.com/questions/5022531/best-way-to-bulk-insert-from-a-c-sharp-datatable

**Comment by ewinnington** [comment]
Hello, I hope this is appropriate to add here, I thought that your research might benefit from this experience and code. 

## Bulk Inserts for Oracle, MSSQL, Postgresql

For the examples, it is assumed that there exists an object **Timeseries** _ts_ with a field **IEnumerable<Datapoint>** **TimeseriesData** that contains Datapoints. A **Datapoint** has a **DateTime** field named _t_ and a **Decimal** field named _value_. 

### Oracle: 

With ODP.NET Unmanaged (Drivers installed seperately)
using Oracle.DataAccess.Client;
using Oracle.DataAccess.Types;

With ODP.NET Managed (Nuget Installed)
using Oracle.ManagedDataAccess.Client;
using Oracle.ManagedDataAccess.Types;

### Datatable to Database 
```
DataTable dt = new DataTable();
dt.Columns.Add("IDTIMESERIES", typeof(int));
dt.Columns.Add("TIMEPOINT", typeof(DateTime));
dt.Columns.Add("VALUE1", typeof(decimal));

int nRowsMaxDt = 0;
//ts being a timeseries object that contains a IEnumberable TimeSeriesData of sp (single points) 
foreach (var sp in ts.TimeSeriesData)
{
	object[] oRowDt = new object[3];
	oRowDt[0] = ts.Id; oRowDt[1] = sp.t; oRowDt[2] = (decimal)sp.value; 
	dt.Rows.Add(oRowDt); nRowsMaxDt = nRowsMaxDt + 1;
}
DumpDataTableToDB("DATATIMESERIES", dt, nRowsMaxDt);
```


```
public void DumpDataTableToDB(string TableName, DataTable dt, int nCount)
{
	using (OracleBulkCopy obc = new OracleBulkCopy(OraCon))
	{
		obc.BatchSize = nCount;
		obc.DestinationTableName = TableName;
		obc.WriteToServer(dt);
		obc.Close();
	}
}
```
						
#### Method: Oracle Array Bind, for Vector Insertion 
This is a better method for insertions on Oracle that does not cause issues with dataguard, using arrays of data. 

```
int[] IDTIMESERIES = new int[nRowsCount];
DateTime[] TIMEPOINT = new DateTime[nRowsCount];
decimal[] VALUE1 = new decimal[nRowsCount];

int i = 0;
foreach (var sp in ts.TimeSeriesData)
{
	IDTIMESERIES[i] = ts.Id; TIMEPOINT[i] = sp.t; VALUE1[i] = sp.value;
	i++;
}
if (nRowsCount > 0)
{
	OracleCommand cmdTimeDump = new OracleCommand("Insert into DATATIMESERIES (IDTIMESERIES, TIMEPOINT, VALUE1) VALUES (:1, :2, :3)", OraCon);
	cmdTimeDump.ArrayBindCount = nRowsCount;
	cmdTimeDump.Parameters.Add(new OracleParameter() { OracleDbType = OracleDbType.Int32, Value = IDINSTRUMENTUNDERLYING });
	cmdTimeDump.Parameters.Add(new OracleParameter() { OracleDbType = OracleDbType.Date, Value = TIMEPOINT });
	cmdTimeDump.Parameters.Add(new OracleParameter() { OracleDbType = OracleDbType.Decimal, Value = VALUE1 });
  
	cmdTimeDump.ExecuteNonQuery();
}

```

**Comment by ewinnington** [comment]
Should I add SQLite? It’s basically the query repeated all inside a transaction.

For MySQL, I’ve not had to interact with it in C#, but I would be curious to see what it would look like.

**Comment by Grauenwolf** [comment]
For MySQL, you have to write it out to a CSV file first. Then the database driver can upload the file.

We'll probably set it up to automatically create and delete the CSV file.

**Comment by Grauenwolf** [comment]
For SQLite there is also the UNION ALL trick.

https://stackoverflow.com/questions/1609637/is-it-possible-to-insert-multiple-rows-at-a-time-in-an-sqlite-database

I'll open a separate ticket for it as "Batch Insert", since it isn't really a bulk insert method. (In Chain, a "batch insert" uses SQL. A "bulk insert" uses non-SQL means such as the MySQL CSV trick.)

**Comment by Grauenwolf** [comment]
All bulk insert options have been implemented as part of v3.1
